src/main/java/io/bittiger/ads/activity/AdsSelection.java

```diff
@@ -0,0 +1,76 @@
+package io.bittiger.ads.activity;
+
+import io.bittiger.ads.model.Ad;
+import net.spy.memcached.MemcachedClient;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Set;
+import java.util.HashSet;
+import java.util.Arrays;
+import java.lang.String;
+
+public class AdsSelection {
+    private static AdsSelection instance = null;
+    
+    protected AdsSelection() {
+    }
+
+    public static AdsSelection getInstance() {
+        if (instance == null) {
+            instance = new AdsSelection();
+        }
+        return instance;
+    }
+
+    public List<Ad> getMatchedAds(String[] keywords) {
+        List<Ad> ads = new ArrayList<Ad>();
+        if (keywords == null || keywords.length == 0) {
+            return new ArrayList<Ad>();
+        }
+
+        Set<Ad> set = new HashSet<Ad>();
+        for (String keyword : keywords) {
+            String invKey = "inv" + keyword;
+            Set<Ad> curr = (Set<Ad>) AdsDao.getInstance().getAds(invKey);
+            if (curr == null) {
+                curr = AdsDao.getInstance().traverseFwdIndex(keyword);
+            }
+            addAds(set, curr, keywords);
+        }
+
+        for (Ad ad : set) {
```

Comment:
I think you can use an override constructor here, 
i.e. List<Ad> ads = new ArrayList<Ad>(set);